clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp

```diff
@@ -339,17 +472,85 @@ static Expected<StringRef> runSPIRVCodeGen(StringRef File, const ArgList &Args,
 Error runSYCLLink(ArrayRef<std::string> Files, const ArgList &Args) {
   llvm::TimeTraceScope TimeScope("SYCL device linking");
 
+  std::mutex ImageMtx;
   LLVMContext C;
 
   // Link all input bitcode files and SYCL device library files, if any.
   auto LinkedFile = linkDeviceCode(Files, Args, C);
   if (!LinkedFile)
     reportError(LinkedFile.takeError());
 
+  auto LinkedModule = getBitcodeModule(*LinkedFile, C);
+  if (!LinkedModule)
+    return LinkedModule.takeError();
+  // sycl-post-link step
+  auto SplitModules = runSYCLSplitModule(std::move(*LinkedModule), Args);
+  if (!SplitModules)
+    reportError(SplitModules.takeError());
+
   // SPIR-V code generation step.
-  auto SPVFile = runSPIRVCodeGen(*LinkedFile, Args, C);
-  if (!SPVFile)
-    return SPVFile.takeError();
+  for (size_t I = 0, E = (*SplitModules).size(); I != E; ++I) {
+    std::string SPVFile(OutputFile);
+    SPVFile.append(utostr(I));
+    auto Err = runSPIRVCodeGen((*SplitModules)[I].ModuleFilePath, Args, SPVFile, C);
+    if (Err)
+      return std::move(Err);
+    (*SplitModules)[I].ModuleFilePath = SPVFile;
+  }
+
+  SmallVector<char, 1024> BinaryData;
+  raw_svector_ostream OS(BinaryData);
```

Comment:
Should we use raw_fd_ostream and write out to a file incrementally instead of holding all the data in memory?